src/shims/foreign_items.rs

```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     /// Emulates calling a foreign item using its name.
     fn emulate_foreign_item_by_name(
         &mut self,
+        def_id: DefId,
         link_name: Symbol,
         abi: Abi,
         args: &[OpTy<'tcx, Tag>],
         dest: &PlaceTy<'tcx, Tag>,
     ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
         let this = self.eval_context_mut();
 
+        let tcx = this.tcx.tcx;
+
+        // First deal with any external C functions in linked .so file
+        // (if any SO file is specified).
+        if this.machine.external_c_so_lib.as_ref().is_some() {
+            if let Some(local_id) = def_id.as_local() {
```

Comment:
This is suspicious, why can we only do this for local declarations?

It probably has to do with you using HIR types when you should be using MIR types.

Reply:
Yes you're right -- using regular MIR types we can use `def_id`!